### PR TITLE
[GDN] Tricked kernels: ungated KKT + fused inference via similarity transform

### PR DESCRIPTION
**Pull request by hypnopump**
## Summary

Integrates the [similarity-transform optimization](https://hypnopump.github.io/post.html?slug=accelerating-gdn-inference) (`N = G·M·G⁻¹`) into FLA's GDN forward and backward paths, following [Simon Veitner's observation](https://veitner.bearblog.dev/simple-math-to-speed-up-gdn-prefill/) (also independently noted in [Comba](https://arxiv.org/abs/2506.02475)).

Instead of computing the gated coupling matrix `N` directly (requiring `BT²` exp ops), we:
1. Compute ungated `M` (skip `BT²` exp ops)
2. Solve `M⁻¹` (same cost)
3. Apply `G`/`G⁻¹` diagonal scaling in the WY step (only `2·BT` exp ops)

**Net savings**: 3968 fewer exp ops per chunk (BT=64).

### Changes

- **`wy_fast.py`** (new): fusemaxxed kernel (inference, A in registers), tricked WY fwd/bwd kernels, ungated kkt+solve wrapper
- **`chunk.py`**: Forward uses ungated kkt+solve + tricked WY; backward uses tricked WY bwd with precomputed w,u for cheap dG; inference auto-dispatches fusemaxxed (T≤8K) vs fused (T>8K); w/u caching for T≥2048

### Benchmark results

All benchmarks on **NVIDIA H100**, B=1, `H·Dh = 2048`, bf16. Baseline: FLA 0.5.0 (commit `f52529e`). Averaged over 3 runs.

#### Forward (no_grad) — auto-dispatch inference

Auto-dispatch picks fusemaxxed (kkt+solve+WY fused, A in registers) for T≤8K, and fused (kkt+solve)+WY for T>8K.

| SeqLen | Dh=128 FLA (ms) | Dh=128 Tricked (ms) | Speedup | Dh=256 FLA (ms) | Dh=256 Tricked (ms) | Speedup |
|-------:|--------:|--------:|--------:|--------:|--------:|--------:|
| 1K | 0.452 | 0.324 | **1.39x** | 0.453 | 0.323 | **1.40x** |
| 2K | 0.454 | 0.324 | **1.40x** | 0.460 | 0.332 | **1.39x** |
| 4K | 0.459 | 0.347 | **1.32x** | 0.538 | 0.428 | **1.26x** |
| 8K | 0.572 | 0.529 | **1.08x** | 0.717 | 0.641 | **1.12x** |
| 16K | 0.899 | 0.790 | **1.14x** | 1.092 | 1.084 | **1.01x** |
| 32K | 1.611 | 1.416 | **1.14x** | 1.974 | 1.971 | **1.00x** |
| 64K | 3.014 | 2.648 | **1.14x** | 3.726 | 3.722 | **1.00x** |
| 128K | 5.844 | 5.128 | **1.14x** | 7.201 | 7.177 | **1.00x** |

#### Forward + Backward

| SeqLen | Dh=128 FLA (ms) | Dh=128 Tricked (ms) | Speedup | Dh=256 FLA (ms) | Dh=256 Tricked (ms) | Speedup |
|-------:|--------:|--------:|--------:|--------:|--------:|--------:|
| 1K | 1.487 | 1.741 | 0.85x | 1.611 | 1.521 | **1.06x** |
| 2K | 1.807 | 1.696 | **1.07x** | 1.582 | 1.449 | **1.09x** |
| 4K | 1.835 | 1.698 | **1.08x** | 1.911 | 1.886 | **1.01x** |
| 8K | 2.121 | 2.049 | **1.04x** | 2.858 | 2.820 | **1.01x** |
| 16K | 3.272 | 3.157 | **1.04x** | 4.773 | 4.727 | **1.01x** |
| 32K | 5.840 | 5.629 | **1.04x** | 8.783 | 8.762 | **1.00x** |
| 64K | 10.992 | 10.524 | **1.04x** | 16.881 | 16.789 | **1.01x** |
| 128K | 21.373 | 20.393 | **1.05x** | 33.133 | 32.878 | **1.01x** |

## Test plan

- [ ] Correctness: verify tricked forward matches FLA baseline (max_diff < 0.1, cosine_sim > 0.999)
- [ ] Correctness: verify tricked backward gradients match FLA baseline
- [ ] Benchmark: reproduce forward speedups on H100
- [ ] Test variable-length sequences (cu_seqlens path)
- [ ] Test with l2norm and initial_state

## Summary by CodeRabbit

* **New Features**
  * Added memory efficiency option for improved performance on sequences with varying lengths
  * Introduced optimized inference path for faster inference without backpropagation

* **Refactor**
  * Streamlined internal computation flow for gated delta rule operations to reduce memory usage and improve backward pass efficiency

### CODE REVIEW

File: fla/ops/gated_delta_rule/chunk.py [hunk 1]
```diff
@@ -97,9 +118,58 @@ def chunk_gated_delta_rule_fwd(
         use_exp2=use_exp2,
         transpose_state_layout=transpose_state_layout,
     )
+
+    T = q.shape[1]
+    cache_wu = not mem_efficient and T >= 2048
```
[the remark this answers is not in the document]
Comment (hypnopump) [reply fla/ops/gated_delta_rule/chunk.py:120]: Fixed in 3f037e2a — aligned to `T >= 2048` in both places.

File: fla/ops/gated_delta_rule/chunk.py [hunk 2]
```diff
@@ -44,13 +54,24 @@ def chunk_gated_delta_rule_fwd(
...
         chunk_indices=chunk_indices,
     )
-    # obtain WY representation. u is actually the new v.
-    # fused kkt + solve_tril + recompute_w_u
-    w, u, A = chunk_gated_delta_rule_fwd_intra(
+
+    # Tricked path: ungated kkt+solve, then tricked WY with G/G_inv scaling
+    A = chunk_gated_delta_rule_fwd_kkt_solve(
+        k=k,
+        v=v,
```
[the remark this answers is not in the document]
Comment (hypnopump) [reply fla/ops/gated_delta_rule/chunk.py:58]: Keeping `v` for API parity with `chunk_gated_delta_rule_fwd_intra` — makes call sites symmetric.

File: fla/ops/gated_delta_rule/chunk.py [hunk 3]
```diff
@@ -345,6 +442,9 @@ def chunk_gated_delta_rule(
         cu_seqlens (torch.LongTensor):
             Cumulative sequence lengths of shape `[N+1]` used for variable-length training,
             consistent with the FlashAttention API.
+        mem_efficient (bool):
+            If False (default), caches w and u tensors from the forward pass (for T >= 2048)
```
[the remark this answers is not in the document]
Comment (hypnopump) [reply fla/ops/gated_delta_rule/chunk.py:451]: Fixed in 3f037e2a — threshold is now consistent (`T >= 2048`) everywhere.

File: fla/ops/gated_delta_rule/chunk_fwd.py [hunk 4]
```diff
@@ -400,3 +400,391 @@ def chunk_gated_delta_rule_fwd_intra(
         use_exp2=use_exp2,
     )
     return w, u, A
+
+
+def chunk_gated_delta_rule_fwd_kkt_solve(
+    k: torch.Tensor,
+    v: torch.Tensor,
```
[the remark this answers is not in the document]
Comment (hypnopump) [reply fla/ops/gated_delta_rule/chunk_fwd.py:410]: Keeping for API parity with `chunk_gated_delta_rule_fwd_intra`.

File: fla/ops/gated_delta_rule/wy_fast.py [hunk 5]
```diff
@@ -346,3 +346,260 @@ def prepare_wy_repr_bwd(
...
+    m_t = o_t < T
+    m_A = (o_t[:, None] > o_t[None, :]) & (m_t[:, None] & m_t)
+    b_dA = tl.where(m_A, b_dA, 0)
+    b_dA = tl.dot(b_dA.to(b_A.dtype), b_A)
+    b_dA = tl.dot(b_A, b_dA.to(b_A.dtype))
+
+    b_M = tl.zeros([BT, BT], dtype=tl.float32)
+    b_dA = tl.where(m_A, -b_dA, 0).to(k.dtype.element_ty)
+
+    tl.debug_barrier()
```
[the remark this answers is not in the document]
Comment (hypnopump) [reply fla/ops/gated_delta_rule/wy_fast.py]: Removed in 3f037e2a.

File: fla/ops/gated_delta_rule/chunk.py [hunk 6]
```diff
@@ -14,8 +14,17 @@
...
+    chunk_gated_delta_rule_fwd_intra,
+    chunk_gated_delta_rule_fwd_kkt_solve,
+    tricked_fused_infer_fwd,
+)
+from fla.ops.gated_delta_rule.wy_fast import (
+    prepare_wy_repr_bwd,
+    recompute_w_u_fwd,
+    tricked_prepare_wy_repr_bwd,
+    tricked_recompute_w_u_fwd,
+)
```
[the remark this answers is not in the document]
Comment (hypnopump) [reply fla/ops/gated_delta_rule/chunk.py:24]: Already cleaned up in the latest revision.

File: fla/ops/gated_delta_rule/chunk.py [hunk 7]
```diff
@@ -97,9 +118,58 @@ def chunk_gated_delta_rule_fwd(
         use_exp2=use_exp2,
         transpose_state_layout=transpose_state_layout,
     )
+
+    T = q.shape[1]
+    cache_wu = not mem_efficient and T >= 2048
+    if cache_wu:
+        return g, o, A, w, u, final_state, initial_state
```
[the remark this answers is not in the document]
Comment (hypnopump) [reply fla/ops/gated_delta_rule/chunk.py:122]: Fixed in 3f037e2a — both use `T >= 2048` now.

File: fla/ops/gated_delta_rule/chunk.py [hunk 8]
```diff
@@ -97,9 +118,58 @@ def chunk_gated_delta_rule_fwd(
...
+        transpose_state_layout=transpose_state_layout,
+    )
+    o = chunk_fwd_o(
+        q=q, k=k, v=v_new, h=h, g=g, scale=scale,
+        cu_seqlens=cu_seqlens,
+        chunk_indices=chunk_indices,
+        use_exp2=use_exp2,
+        transpose_state_layout=transpose_state_layout,
+    )
+    return g, o, final_state
```
[the remark this answers is not in the document]
Comment (hypnopump) [reply fla/ops/gated_delta_rule/chunk.py:167]: Fixed in 3f037e2a — no-grad path now requires `cp_context is None`, falls back to full path otherwise.

File: fla/ops/gated_delta_rule/chunk.py [hunk 9]
```diff
@@ -97,9 +115,58 @@ def chunk_gated_delta_rule_fwd(
         use_exp2=use_exp2,
         transpose_state_layout=transpose_state_layout,
     )
+
+    T = q.shape[1]
+    cache_wu = not mem_efficient and T >= 2048
+    if cache_wu:
+        return g, o, A, w, u, final_state, initial_state
     return g, o, A, final_state, initial_state
```
[the remark this answers is not in the document]
Comment (hypnopump) [reply fla/ops/gated_delta_rule/chunk.py:123]: Fixed in 3f037e2a.

File: fla/ops/gated_delta_rule/chunk.py [hunk 10]
```diff
@@ -408,6 +505,30 @@ def chunk_gated_delta_rule(
...
+            g=g,
+            beta=beta,
+            scale=scale,
+            initial_state=initial_state,
+            output_final_state=output_final_state,
+            cu_seqlens=cu_seqlens,
+            chunk_indices=chunk_indices,
+            transpose_state_layout=transpose_state_layout,
+        )
+        return o.to(q.dtype), final_state
```
[the remark this answers is not in the document]
Comment (hypnopump) [reply fla/ops/gated_delta_rule/chunk.py]: Fixed in 3f037e2a — guarded with `cp_context is None`.

File: fla/ops/gated_delta_rule/chunk.py [hunk 11]
```diff
@@ -263,12 +338,24 @@ def forward(
             cp_context=cp_context,
             chunk_indices=chunk_indices,
             transpose_state_layout=transpose_state_layout,
+            mem_efficient=mem_efficient,
         )
-        ctx.save_for_backward(q, q_rstd, k, k_rstd, v, g, beta, A, initial_state, cu_seqlens, chunk_indices)
+
+        T = q.shape[1]
+        cache_wu = not mem_efficient and T > 2048
+        if cache_wu:
```
[the remark this answers is not in the document]
Comment (hypnopump) [reply fla/ops/gated_delta_rule/chunk.py:346]: Fixed in 3f037e2a — changed to `T >= 2048`.

File: fla/ops/gated_delta_rule/wy_fast.py [hunk 12]
```diff
@@ -346,3 +346,264 @@ def prepare_wy_repr_bwd(
...
+        b_vb = (b_v * (b_b * b_G_inv)[:, None]).to(b_v.dtype)
+        b_u = tl.dot(b_A, b_vb, allow_tf32=False) * b_G[:, None]
+        tl.store(p_u, b_u.to(p_u.dtype.element_ty), boundary_check=(0, 1))
+
+    for i_k in range(tl.cdiv(K, BK)):
+        p_k = tl.make_block_ptr(k + (bos*H + i_h) * K, (T, K), (H*K, 1), (i_t * BT, i_k * BK), (BT, BK), (1, 0))
+        p_w = tl.make_block_ptr(w + (bos*H + i_h) * K, (T, K), (H*K, 1), (i_t * BT, i_k * BK), (BT, BK), (1, 0))
+        b_k = tl.load(p_k, boundary_check=(0, 1))
+        b_w = tl.dot(b_A, (b_k * b_b[:, None]).to(b_k.dtype)) * b_G[:, None]
+        tl.store(p_w, b_w.to(p_w.dtype.element_ty), boundary_check=(0, 1))
```
[the remark this answers is not in the document]
Comment (hypnopump) [reply fla/ops/gated_delta_rule/wy_fast.py:426]: Good catch — fixed in 19145b96. The `g_max` normalization only cancels for `u` (where `G * G_inv = 1`), but `w = G * (A @ k*β)` has no inverse term. Now using `G_full = G * exp(g_max)` for the w path in all three kernels (fwd, bwd, fusemaxxed).
